### PR TITLE
[CIRC-408] Implement sending time-based not-real-time loan notices with multiple items

### PR DESCRIPTION
**Pull request by dmytrotur**
## Purpose
Implement sending time-based not-real-time loan notices with multiple items
Resolves: [CIRC-408](https://issues.folio.org/browse/CIRC-408)

## Approach

- Add new timer endpoint for processing due date not-real-time scheduled notices
- Fetch scheduled notices ordered by **recipientId**, **templateId**, **triggeringEvent**, **timing**, **format** to be able to group them

## Pre-Merge Checklist:
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [x] Code coverage on new code is 80% or greater
  - [x] Duplications on new code is 3% or less
  - [x] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] Were any API paths or methods changed, added or removed?
  - [ ] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [ ] Were permissions changed, added, or removed?
  - [ ] Are there new interface dependencies?
  - [x] There are no breaking changes in this PR.

### CODE REVIEW

File: src/main/java/org/folio/circulation/domain/notice/NoticeFormat.java [hunk 1]
```diff
@@ -5,6 +5,8 @@
 public enum NoticeFormat {
 
   EMAIL("Email", "email", "text/html"),
+  PRINT("Print", "print", "text/html"),
```
Comment (marcjohnson-kint) [inline src/main/java/org/folio/circulation/domain/notice/NoticeFormat.java]: My understanding is that this change is for batching email notices together. Why are these additional formats needed for this?

Comment (dmytrotur) [reply src/main/java/org/folio/circulation/domain/notice/NoticeFormat.java]: This is correct. I added them just to match circulation-storage schema, where `Print` and `SMS` are allowed values. Would you like to remove these changes as they are not necessary now?

Comment (marcjohnson-kint) [reply src/main/java/org/folio/circulation/domain/notice/NoticeFormat.java]: If they are going to be treated the same as unknown, I don't really understand the value in having them.

If it makes it easier to understand the mapping between the two contexts (storage and business logic) then I think they can remain.

I'm trying to ask questions about why changes are made to understand the decisions better and to encourage reflection, for something like this, I leave the decision to you.

Comment (dmytrotur) [reply src/main/java/org/folio/circulation/domain/notice/NoticeFormat.java]: Given these formats are not supported now, it seems like there is no point in them.  ISo I'm going to remove them for now.

File: src/main/java/org/folio/circulation/support/CqlOrder.java [hunk 2]
```diff
@@ -0,0 +1,24 @@
+package org.folio.circulation.support;
+
+public class CqlOrder {
```
Comment (marcjohnson-kint) [inline src/main/java/org/folio/circulation/support/CqlOrder.java]:
```suggestion
public class CqlSortOrder {
```

File: src/main/java/org/folio/circulation/support/CqlSortDirection.java [hunk 3]
```diff
@@ -0,0 +1,16 @@
+package org.folio.circulation.support;
+
+public enum CqlSortDirection {
+
+  ASC("sort.ascending"), DESC("sort.descending");
```
Comment (marcjohnson-kint) [inline src/main/java/org/folio/circulation/support/CqlSortDirection.java]: Personally, I'd be tempted to use the full words here. Is there a reason to keep them short?

File: src/main/java/org/folio/circulation/support/CqlSortBy.java [hunk 4]
```diff
@@ -1,51 +1,48 @@
...
+  public static CqlSortBy descending(String index) {
+    return CqlSortBy.sortBy(CqlOrder.desc(index));
+  }
 
-    @Override
-    protected String applyTo(String query) {
-      return String.format("%s sortBy %s/sort.ascending", query, index);
-    }
+  public static CqlSortBy none() {
+    return CqlSortBy.sortBy(Collections.emptyList());
```
Comment (marcjohnson-kint) [inline src/main/java/org/folio/circulation/support/CqlSortBy.java]: Why not use the private constructor here?

File: src/main/java/org/folio/circulation/support/CqlOrder.java [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+
+  private final String index;
+  private final CqlSortDirection direction;
+
+  public CqlOrder(String index, CqlSortDirection direction) {
+    this.index = index;
+    this.direction = direction;
+  }
+
+  public static CqlOrder asc(String index) {
```
Comment (marcjohnson-kint) [inline src/main/java/org/folio/circulation/support/CqlOrder.java]: Personally, I'd be tempted to use the full words here. Is there a reason to keep them short?

File: src/test/java/api/support/fixtures/ScheduledNoticeProcessingClient.java [hunk 6]
```diff
@@ -10,25 +10,39 @@
...
-    DateTimeUtils.setCurrentMillisSystem();
+    wrapWithTimeMocking(this::runRequestNoticesProcessing, mockSystemTime);
   }
 
   public void runRequestNoticesProcessing() {
     URL url = circulationModuleUrl("/circulation/request-scheduled-notices-processing");
     manuallyStartTimedTask(url, 204, "request-scheduled-notices-processing-request");
   }
+
+  private void wrapWithTimeMocking(Runnable runnable, DateTime mockSystemTime) {
```
Comment (marcjohnson-kint) [inline src/test/java/api/support/fixtures/ScheduledNoticeProcessingClient.java]:
```suggestion
  private void runWithFrozenTime(Runnable runnable, DateTime mockSystemTime) {
```

File: src/test/java/api/support/fakes/FakeCQLToJSONInterpreter.java [hunk 7]
```diff
@@ -40,7 +41,16 @@ public FakeCQLToJSONInterpreter(boolean diagnosticsEnabled) {
     }
   }
 
-  private Comparator<? super JsonObject> sortForQuery(String sort) {
+  private Comparator<JsonObject> sortForQuery(String sort) {
```
Comment (marcjohnson-kint) [inline src/test/java/api/support/fakes/FakeCQLToJSONInterpreter.java]: Please can a test to demonstrate that the fake can sort by multiple clauses

File: src/main/java/org/folio/circulation/support/CqlOrder.java (hunk 2 above)
Comment (dmytrotur) [reply src/main/java/org/folio/circulation/support/CqlOrder.java]: Done

File: src/main/java/org/folio/circulation/support/CqlSortDirection.java (hunk 3 above)
Comment (dmytrotur) [reply src/main/java/org/folio/circulation/support/CqlSortDirection.java]: Renamed to full words

File: src/main/java/org/folio/circulation/support/CqlOrder.java (hunk 5 above)
Comment (dmytrotur) [reply src/main/java/org/folio/circulation/support/CqlOrder.java]: Renamed to full words

File: src/main/java/org/folio/circulation/support/CqlSortBy.java (hunk 4 above)
Comment (dmytrotur) [reply src/main/java/org/folio/circulation/support/CqlSortBy.java]: Changed

File: src/test/java/api/support/fixtures/ScheduledNoticeProcessingClient.java (hunk 6 above)
Comment (dmytrotur) [reply src/test/java/api/support/fixtures/ScheduledNoticeProcessingClient.java]: Done

File: src/test/java/api/support/fakes/FakeCQLToJSONInterpreter.java (hunk 7 above)
Comment (dmytrotur) [reply src/test/java/api/support/fakes/FakeCQLToJSONInterpreter.java]: Done

File: src/main/java/org/folio/circulation/domain/notice/schedule/ScheduledNoticeGroupDefinition.java [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+    this.noticeFormat = noticeFormat;
+    this.noticeTiming = noticeTiming;
+  }
+
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) return true;
+
+    if (o == null || getClass() != o.getClass()) return false;
+
```
Comment (bohdan-suprun) [inline src/main/java/org/folio/circulation/domain/notice/schedule/ScheduledNoticeGroupDefinition.java]: Hi @dmytrotur, could you please fix checkstyle here:
 ```suggestion
if (this == o) {
 return true;
}
if (o == null || getClass() != o.getClass()) { 
 return false;
}
```

Comment (dmytrotur) [reply src/main/java/org/folio/circulation/domain/notice/schedule/ScheduledNoticeGroupDefinition.java]: @bohdan-suprun Done.

File: src/main/java/org/folio/circulation/resources/DueDateNotRealTimeScheduledNoticeProcessingResource.java [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+
+  public DueDateNotRealTimeScheduledNoticeProcessingResource(HttpClient client) {
+    super("/circulation/due-date-not-real-time-scheduled-notices-processing", client);
+  }
+
+  @Override
+  protected CompletableFuture<Result<MultipleRecords<ScheduledNotice>>> findNoticesToSend(
+    ScheduledNoticesRepository scheduledNoticesRepository, int limit) {
+
+    List<CqlSortClause> cqlSortClauses = Stream.of(
```
Comment (bohdan-suprun) [inline src/main/java/org/folio/circulation/resources/DueDateNotRealTimeScheduledNoticeProcessingResource.java]: @dmytrotur, can we declare the list as a class field or constant?

File: src/main/java/org/folio/circulation/resources/DueDateNotRealTimeScheduledNoticeProcessingResource.java [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+      : max(orderedGroups.size() - 1, 1);
+
+    List<Pair<ScheduledNoticeGroupDefinition, List<ScheduledNotice>>> noticeGroups =
+      orderedGroups.entrySet()
+        .stream()
+        .limit(limit)
+        .map(e -> Pair.of(e.getKey(), e.getValue()))
+        .collect(Collectors.toList());
+
+    return dueDateNoticeHandler.handleNotices(noticeGroups)
```
Comment (bohdan-suprun) [inline src/main/java/org/folio/circulation/resources/DueDateNotRealTimeScheduledNoticeProcessingResource.java]: @dmytrotur, why we cannot use map here, and using list of pairs instead?

Comment (dmytrotur) [reply src/main/java/org/folio/circulation/resources/DueDateNotRealTimeScheduledNoticeProcessingResource.java]: @bohdan-suprun Thanks for pointing that. I just realized we don't need a list of pairs here, it enough to have a list of lists.

Comment (dmytrotur) [reply src/main/java/org/folio/circulation/resources/DueDateNotRealTimeScheduledNoticeProcessingResource.java]: Done

File: src/main/java/org/folio/circulation/resources/DueDateNotRealTimeScheduledNoticeProcessingResource.java (hunk 9 above)
Comment (dmytrotur) [reply src/main/java/org/folio/circulation/resources/DueDateNotRealTimeScheduledNoticeProcessingResource.java]: Moved to constant

File: src/main/java/org/folio/circulation/resources/DueDateNotRealTimeScheduledNoticeProcessingResource.java (hunk 10 above)
Comment (bohdan-suprun) [reply src/main/java/org/folio/circulation/resources/DueDateNotRealTimeScheduledNoticeProcessingResource.java]: Thanks @dmytrotur

**Pull request merged**